zerver/views/auth.py

```diff
@@ -543,8 +555,73 @@ def update_login_page_context(request: HttpRequest, context: Dict[str, Any]) ->
 
     context['wrong_subdomain_error'] = WRONG_SUBDOMAIN_ERROR
 
+class PatchedExtraSessionStorage(ExtraSessionStorage):
+    def reset(self) -> None:
+        if self.prefix in self.request.session:
+            super(ExtraSessionStorage, self).reset()
+        else:
+            self.init_data()
+
+class TwoFactorLoginView(BaseTwoFactorLoginView):
+    extra_context = None  # type: ExtraContext
+    storage_name = 'zerver.views.auth.PatchedExtraSessionStorage'
+    form_list = (
+        ('auth', OurAuthenticationForm),
+        ('token', AuthenticationTokenForm),
+        ('backup', BackupTokenForm),
+    )
+
+    def __init__(self, extra_context: ExtraContext=None,
+                 *args: Any, **kwargs: Any) -> None:
+        self.extra_context = extra_context
+        super().__init__(*args, **kwargs)
+
+    def get_form_kwargs(self, step: str) -> Dict[str, Any]:
+        kwargs = super(TwoFactorLoginView, self).get_form_kwargs(step)
+        if step == 'auth':
+            kwargs.update({'request': self.request})
+        return kwargs
```

Comment:
I don't understand why this override does anything; doesn't this already set `request` that way?
```
    def get_form_kwargs(self, step=None):                                                            
        """                                                                                          
        AuthenticationTokenForm requires the user kwarg.                                             
        """                                                                                          
        if step == 'auth':                                                                           
            return {                                                                                 
                'request': self.request                                                              
            }                                                                                        
        if step in ('token', 'backup'):                                                              
            return {                                                                                 
                'user': self.get_user(),                                                             
                'initial_device': self.get_device(step),                                             
            }                                                                                        
        return {}                                                                                    
```